src/coreclr/vm/gcinfodecoder.cpp

```diff
@@ -299,22 +239,118 @@ GcInfoDecoder::GcInfoDecoder(
     if (remainingFlags == 0)
     {
         // Bail, if we've decoded enough,
-        return;
+        return true;
     }
 
 #ifdef FIXED_STACK_PARAMETER_SCRATCH_AREA
-    if (slimHeader)
+    m_SizeOfStackOutgoingAndScratchArea = (UINT32)DENORMALIZE_SIZE_OF_STACK_AREA(m_Reader.DecodeVarLengthUnsigned(SIZE_OF_STACK_AREA_ENCBASE));
+#endif // FIXED_STACK_PARAMETER_SCRATCH_AREA
+
+    return false;
+}
+
+GcInfoDecoder::GcInfoDecoder(
+            GCInfoToken gcInfoToken,
+            GcInfoDecoderFlags flags,
+            UINT32 breakOffset
+            )
+            : m_Reader(dac_cast<PTR_CBYTE>(gcInfoToken.Info))
+            , m_InstructionOffset(breakOffset)
+            , m_IsInterruptible(false)
+            , m_ReturnKind(RT_Illegal)
+#ifdef _DEBUG
+            , m_Flags( flags )
+            , m_GcInfoAddress(dac_cast<PTR_CBYTE>(gcInfoToken.Info))
+#endif
+           , m_Version(gcInfoToken.Version)
+{
+    _ASSERTE( (flags & (DECODE_INTERRUPTIBILITY | DECODE_GC_LIFETIMES)) || (0 == breakOffset) );
+
+    // The current implementation doesn't support the two flags together
+    _ASSERTE(
+        ((flags & (DECODE_INTERRUPTIBILITY | DECODE_GC_LIFETIMES)) != (DECODE_INTERRUPTIBILITY | DECODE_GC_LIFETIMES))
+            );
+
+    //--------------------------------------------
+    // Pre-decode information
+    //--------------------------------------------
+
+    bool slimHeader = (m_Reader.ReadOneFast() == 0);
+    // Use flag mask to bail out early if we already decoded all the pieces that caller requested
+    int remainingFlags = flags == DECODE_EVERYTHING ? ~0 : flags;
+
+    if (!slimHeader)
```

Comment:
Predecoding a fat header could be much more involved compared to slim headers, so let's deal with fat headers in a separate helper.